### DISCUSSION TITLE
project management: pinning versions

### DISCUSSION THREAD

**Discussion by pwoolvett** [flakeheaven/flakeheaven#34] [Governance]
I think we should at least specify the major version for semver packages int their `pyproject.toml` entries. This way, we avoid flakeheaven breaking for new users when major changes occur within deps.

We might be more agressive with other styles (calver? etc) and pin the whole version...

dependabot could help keeping libs updated :)

What do you think?

**Comment by jose-rojas0** [comment]
this sounds good to me 👍 specifying major version seems like common practice in other open source libs

**Comment by pwoolvett** [comment]
PRs are welcome! see #55

**Comment by lyz-code** [comment]
Hey, I might disagree. I'd rather not introduce any upper caps on any library (even major), [Henry Schreiner explained deeply why](https://iscinumpy.dev/post/bound-version-constraints/), it introduces a lot of problems in exchange of a false sense of security.

That's also the reason why I wouldn't use `poetry`, as they do upper capping by default. We can evaluate `pdm`,  I did a [summary on differences](https://lyz-code.github.io/blue-book/python_package_management/) between `poetry`, `pdm` and `pipenv`

**Reply by pwoolvett** [reply]
I see you've given it  a lot of thought. I would add there should be explicit distinctions between applications, apis, and libraries. In this scenario, I think flakeheaven is located towards the "only application" side, as we are not exposing api (or at least suggesting it by underscoring every single module :laughing: ). See, for example, [docker-compose](https://github.com/docker/compose/issues/4542#issuecomment-283191533).

Having said that, my vote goes in favor of whichever has the less extra work for maintainers. In the same line, I would like to also enforce conventional commits and automatic release +changelog, see [the other discussion](https://github.com/flakeheaven/flakeheaven/discussions/50).

The way I see it, we either

A) **pin upper major** and let dependabot / user requests upgrade our deps and test using those (asumming, imperfect as it is, semver).
  - PRO: flakeheaven breaks less often for new setups
  - CON: produce unnecessary, "ficticious" incompatibilities such as (probably( [your comment](https://github.com/pdm-project/pdm/issues/908))
  - WORK: accept prs from dependabot

B) **pin lower major** and explicitly block updates ONCE PROBLEMS ARISE.
  - PRO: less ficticious incompatibilities
  - CON: would break more often for new setups
  - WORK: iff something breaks, we either (i) pin major,  (ii) refactor our code to use new api, (iii) ignore it and refer to the poetry lock
  
  @lyz-code would you care to complement here your thoughts on the summary

**Reply by pwoolvett** [reply]
regarding the major capping by default with poetry, I wouldn't consider it an issue as its easy to use `poetry add "package<version-spec>`" instead of just `poetry add pacakge`

**Reply by lyz-code** [reply]
Thanks for the quick answer @pwoolvett, indeed I've given it much thought after spending most of my repository maintenance time on opening PR on the downstream dependencies to change/remove their upper caps and my CIs being broken. Even though I've got a strong opinion on it, I'd support with whatever we all decide :).

>  In this scenario, I think flakeheaven is located towards the "only application" side, as we are not exposing api

I'd agree with you if `flakeheaven` was used as a standalone api, but given that it's a tool meant to help on the development, I'd treat it as a library, as it's very probable that the developers add them under their development dependencies therefore being part of the dependency conflict resolution.

> I would like to also enforce conventional commits and automatic release +changelog

+1000 to this, I've also written about this:

* [On program versioning](https://lyz-code.github.io/blue-book/versioning/#deciding-what-version-system-to-use-for-your-programs)
* [On the changelog](https://lyz-code.github.io/blue-book/changelog/)

> A) pin upper major and let dependabot / user requests upgrade our deps and test using those (asumming, imperfect as it is, semver).

I'd vote against this scenario based on the arguments of [Henry's article](https://iscinumpy.dev/post/bound-version-constraints/), and all the time I've spent fixing CIs due to projects following this strategy. I don't want to rewrite the whole article here :P, so please read (or skim through it) to get an idea of my reasons. If it's too long, I've [made a summary](https://lyz-code.github.io/blue-book/versioning/)

> B) pin lower major and explicitly block updates ONCE PROBLEMS ARISE.

Instead of blocking the lower major, I'd pin the current version of the libraries with `>=` as it's the version of the dependency your tests make sure it works, you don't know if an earlier version breaks your code.

Instead of relying on dependabot to do PRs or users to make requests, an automatic CI should everyday update the requirements, run the tests, and if they pass update the `poetry.lock` (or similar lock), that way you don't upper cap, and make sure that your code works with the newest versions of downstream libraries. If the CI doesn't pass we'll then debug where does the problem come from, open the required issues on downstream repos, refactor the code or block the specific version of the library. 

This is the workflow I've been using to maintain my code, you can see an example [here](https://github.com/lyz-code/autoimport/actions) ([code](https://github.com/lyz-code/autoimport/tree/master/.github/workflows)), and it has proven to be the one that has needed the less work on my side.

> I wouldn't consider it an issue as its easy to use `poetry add "package<version-spec>"` instead of just `poetry add package`

The workflow wouldn't be that, you'd first need to search which is the latest version of the package and then write it in the `poetry add` not undoable, but not ideal either. It's true that `pdm` is yet less popular, but it supports this by default, you define the update strategy and then you can do `pdm add package` and everything works smoothly.

In my experience `pdm` behaves much like `poetry` with additional features, the only downside that I've found is that dependabot doesn't yet support `pdm.lock` but if we follow the update workflow I proposed that wouldn't be a problem.

To get more info on `pdm` vs `poetry` I'd read these references:

* [PDM developer comparison](https://dev.to/frostming/a-review-pipenv-vs-poetry-vs-pdm-39b4)
* [John Franey comparison](https://johnfraney.ca/posts/2019/03/06/pipenv-poetry-benchmarks-ergonomics/)
* [Frost Ming analsys (developer of PDM)](https://frostming.com/2021/03-26/pm-review-2021/#result)
* [Henry Schreiner analysis on Poetry](https://iscinumpy.dev/post/poetry-versions/)

**Comment by lyz-code** [comment]
One example of upper pinning versions: I'm unable to [add flakeheaven with pdm](https://github.com/pdm-project/pdm/issues/908) due to the upper cap on `importlib-metadata`

**Comment by pwoolvett** [comment]
Thanks for the feedback @lyz-code .

I wouldn't mind to try your proposed solutions, besides (1) sticking to semver  and (2) poetry, at least for the moment.

any other thougts @jarojasm95 @mcarans ?

I think once we've settled this, we have to automate the publishing ASAP.

**Reply by mcarans** [reply]
For some reason, I missed the notification about this discussion. Sorry about that. 

I am using an older workflow than either poetry or pdm, so cannot comment on their pros and cons. For the packages I publish, I pin dependencies in requirements.txt for testing, but use >= for any dependencies in setup.cfg where I've found I need a bug fix or feature in a specific version. I've run across a problem with a library that upper pinned before so I don't do that. If there is a problem, I fix it and then add or change the dependencies with >= as appropriate.

I agree about the automation of publishing. If it is done by a release in GitHub kicking off publishing to PyPI (something like [this](https://github.com/OCHA-DAP/hdx-python-api/blob/main/.github/workflows/publish.yml)), that will resolve this sort of [issue](https://github.com/flakeheaven/flakeheaven/issues/63) (tagging of released versions).
